src/js/builtins/BundlerPlugin.ts

```diff
@@ -511,6 +543,48 @@ export function runOnLoadPlugins(
 
   const generateDefer = () => this.generateDeferPromise(internalID);
   var promiseResult = (async (internalID, path, namespace, isServerSide, defaultLoader, generateDefer) => {
+    // Check for virtual modules first (file namespace and in virtualModules map)
+    if (this.virtualModules && this.virtualModules.$has(path) && (namespace === "file" || namespace === "")) {
+      const virtualModuleCallback = this.virtualModules.$get(path);
+      if (virtualModuleCallback) {
+        let result;
+        try {
+          result = virtualModuleCallback();
+        } catch (e) {
+          // If the callback throws, report it as an error
+          this.addError(internalID, e, 1);
+          return null;
+        }
+
+        try {
+          if (!result || !$isObject(result)) {
+            throw new TypeError('Virtual module must return an object with "contents" property');
+          }
+
+          var { contents, loader = defaultLoader } = result;
+
+          if (!(typeof contents === "string")) {
+            throw new TypeError('Virtual module must return an object with "contents" as a string');
+          }
+
+          if (!(typeof loader === "string")) {
+            throw new TypeError('Virtual module "loader" must be a string if provided');
```

Comment:
Should include the module name in these error messages? So developers can more easily know which module was broken